modules/asr/Cyberon/DSpotter_asr.c

```diff
@@ -0,0 +1,206 @@
+    oDSpotterInitData.nCommandStageFlowControl = COMMAND_STAGE_MULTI_COMMAND;
+    oDSpotterInitData.byAGC_Gain = (uint8_t)(VOLUME_SCALE_RECONG / 100);
+    oDSpotterInitData.bOneShotMode = true;
+    oDSpotterInitData.bExternalFlashModel = true;
+
+    nMemSize = DSpotterHL_GetMemoryUsage((const uint32_t *)model, &oDSpotterInitData);
+    DBG_TRACE("The DSpotter memory usage is %d.\r\n", nMemSize);
+    g_lpbyDSpotterMem = devmem_malloc(devmem_ctx, nMemSize);
+    if (g_lpbyDSpotterMem == NULL)
+    {
+        DBG_TRACE("devmem_malloc() fail!\r\n");
+        return NULL;
+    }
+
+    DBG_TRACE("DSpotterHL_Init\r\n");
+    nRet = DSpotterHL_Init((const uint32_t *)model, &oDSpotterInitData, g_lpbyDSpotterMem, nMemSize);
+    if (nRet != DSPOTTER_SUCCESS)
+    {
+        DBG_TRACE("DSpotterHL_Init() fail, error = %d!\r\n", nRet);
+        devmem_free(devmem_ctx, g_lpbyDSpotterMem);
+        g_lpbyDSpotterMem = NULL;
+        return NULL;
+    }
+
+    DBG_TRACE("The list of trigger word: \r\n");
+    nCount = DSpotterHL_GetDisplayCommandCount(DSPOTTER_HL_TRIGGER_STAGE);
+    for (int i = 0; i < nCount; i++)
+    {
+        DSpotterHL_GetDisplayCommand(DSPOTTER_HL_TRIGGER_STAGE, i, szCommand, sizeof(szCommand), &nCmdID);
+        DBG_TRACE("    %s, ID = %d\r\n", szCommand, nCmdID);
+    }
+
+    nCount = DSpotterHL_GetDisplayCommandCount(DSPOTTER_HL_COMMAND_STAGE);
+    if (nCount > 0)
+    {
+        DBG_TRACE("The list of command word: \r\n");
+        for (int i = 0; i < nCount; i++)
+        {
+            DSpotterHL_GetDisplayCommand(DSPOTTER_HL_COMMAND_STAGE, i, szCommand, sizeof(szCommand), &nCmdID);
+            DBG_TRACE("    %s, ID = %d\r\n", szCommand, nCmdID);
+        }
+    }
+    DBG_TRACE("\r\n");
+
+    return (asr_port_t)100;
+}
+
+asr_error_t asr_get_attributes(asr_port_t *ctx, asr_attributes_t *attributes)
+{
+    return ASR_NOT_SUPPORTED;
+}
+
+//static int max_duration = 0;
+
+asr_error_t asr_process(asr_port_t *ctx, int16_t *audio_buf, size_t buf_len)
+{
+#ifdef UART_DUMP_RECORD
+    uint8_t byaTxBuffer[DSPOTTER_FRAME_SAMPLE*sizeof(int16_t)*3/2];
+    int nTransferSize;
+
+    if (buf_len < DSPOTTER_FRAME_SAMPLE)
+    {
+        nTransferSize = Convert2TransferBuffer((const uint8_t*)audio_buf, buf_len*sizeof(int16_t), byaTxBuffer, sizeof(byaTxBuffer), eFourByteDataOneChecksum);
+        rtos_uart_tx_write(uart_tx_ctx, byaTxBuffer, (uint32_t)nTransferSize);
+    }
+#else
+    if (++g_nRecordFrameCount % 100 == 0)
+        DBG_TRACE(".");
+#endif
+
+#ifdef SKIP_DSPOTTER_RECOG
+    return ASR_ERROR;
+#endif
+    // uint32_t timer_start = get_reference_time();
+
+    int nRet = DSpotterHL_AddSampleNoFlow(audio_buf, buf_len);
+
+    // uint32_t timer_end = get_reference_time();
+    // uint32_t duration = timer_end - timer_start;
+    // if (duration>max_duration) {
+    // DBG_TRACE("Duration: %lu (us)\n", duration / 100);
+    // max_duration = duration;
+    // }
```

Comment:
There is quite a bit of commented out code in this file.  There may be a good reason for it to stay, but please either add a comment stating that reason or remove it.

Reply:
I added some comments